pkgs/moduleit/module-definition.nix

```diff
@@ -580,7 +580,7 @@ in
       let
 
         moduleJSON = {
-          id = config.id;
+          id = "${config.id}-v${config.version}";
```

Comment:
This is going to be confusing because the id people specify doesn't match the id used in the json.

There is something similar to this in nixpkgs. They have a derivation "name", which is defined as `"${pname}-${version}"`.

We either should use a different name for one of these, or figure out how to get rid of one of them.

Reply:
@masad-frost favors removing version altogether (I guess move to the version is implicitly inside the ID). But I think that'd be weird for author to have to change their package's ID each time before they publish an update. Hmm, not sure...

Reply:
Maybe call it `registryId`; or just revert this change - you can always concatenate the id + version to get the same thing.

Reply:
I have a strong opinion that we shouldn't have 2 things we refer to as version and one of them has no real versioning concept/is not part of the versioning system.

Reply:
@masad-frost I think you are saying the upgrade maps system removes the need for the version field?

Although the upgrade maps implements a versioning system, I don't think of it as overlapping with the term "version", because I associate the word with numeric versions or some sort of tag. Even when using the maps, we'd still need to use a versioning scheme in addition to it, whether numeric, semver-based, date-based, or sha-based. It's true that we could allow each author to independently choose their own versioning scheme.

I'd like to think through some alternatives from the authoring perspective. But first, I'll summarize the current design.

## Current Design

The IDs within the registry (let's call this the *registry module ID*) look like:

`nodejs-18-v1.2` which is composed of `<module ID>-v<version>`

In this case the "community version" is embedded in the module ID, but that's up to the module author, and it's no longer part of our system. The author writes something like:

```
{
  id = "nodejs-18";
  name = "Node.js 18 Tools";
  version = "1.2";
  ...
}
```

We have a lock_modules.py script that pins each unique module registry ID to a Nix output path. If a change is detected, the author is forced to increment the version number before publishing. (Is this script necessary? More below...)

## Alternative 1

Having both id and name is probably redundant and we can probably have just name. So the
author would write:

```
{
  name = "nodejs-18";
  version = "1.2";
  ...
}
```

An ID field can be generated using `"${name}-v${version}"`, and we can use the "description" field for more wordy user-facing description of the module.

*Pro:* familiar to users of popular packages like npm

## Alternative 2

If we remove the version field altogether, it puts the onus on the user to choose a versioning scheme:

```
{
  id = "nodejs-18-1.2";
}
```

or


```
{
  id = "nodejs-18-20230518";
}
```

or

```
{
  id = "nodejs-18-8ca571d3efa03de178c4262c5b908ef1e3ff2218";
}
```

Substitute "name" for "id" if you prefer.

Again, a lock_modules.py script can be used to pin each registry ID to an output path. If something changed, the user would be forced to change the ID.

We may want to expose the upgrade map system to users at some point, so it'd serve them well to use a scheme that's easy to work with when adding entries to and editing the map.

*Pro:* more user choice; less structure in our system
*Con:* less uniformity among modules

## Alterative 3

We could chose to just use git to generate the version tags, in which case the user will not need to concern themselves with updating the version when they change code:

```
{
  id = "nodejs-18";
}
```

The lock_modules.py script would generate registry IDs like: `nodejs-18-<commit sha>`. Or it can be date-based: `nodejs-18-202305180934`, based on the timestamp of the commit.

*Pro:* no need to maintain version field or id field
*Con:* the commit sha-based version scheme would be hard to work with when editing the upgrade maps.
Date-based would be better, but not as easy as a numeric scheme.

## Do we need the lock_modules.py script?

This script gives these things:

1. listing the build inputs of all historical versions which is needed by the current Nix modules disk scheme to build them and put them on the disk
2. the same listing is used in the module/extensions store UI
3. ensures the same registry module ID cannot generate multiple different Nix output paths in different versions of the repo

It comes with these problems:
1. developer maintenance overhead (update / merging)
2. it's an additional layer. Do we push it onto community authors as well?

we can maybe get away with not using this script:

* 1 and 2 can be done by go through every historical commit in the repo and evaluating (not building) every module derivation to
get all possible `<module ID>-<commit ID>` -> Nix output path mappings. This could be expensive if the repo has many commits and/or is large in size.
* 3 becames unnecessary
* alternative 3 lends itself better to this approach because conflicting entries is not possible

## Community Modules

We haven't talked about how to handle updates (auto and recommended) with community Nix modules, but maybe that should factor into the design here
* should they be able to specify upgrade maps?
* how to list their modules in the store?
* to they open a PR in our repo or publish their own repo?
* can users just put the git URL of a 3rd party repo + build target in the `modules` array in `.replit`?

## Other Alternatives?

If you think of other alternatives, please share.

## Analysis

Additional pros/cons of the current design vs each alternative?

Reply:
One more thing to throw it for consideration: using upgrade maps for auto-upgrading minor and patch versions could get tedious if there are frequent updates to a module, and we'd also end up with many entries in the map. We can consider using upgrade maps only for recommended upgrade, but a semver-based for auto upgrades. Or coming up with automation to add entries to the map for the auto upgrade case.

Reply:
Had a discussion and agreement last Friday, but to have this in written form:

1. We decided on going with alternative 3
2. And we are keeping lock_modules.py, which will take on the job of automatically generating module registry IDs in the form of `v<version>-<date>-<commit>` where:
    * `version` is a numeric auto-incremented number that starts at 1
    * `date` is an 8-digit sequence of the form: `YYYYMMDD`
    * `commit` is a 7-digit prefix of the git commit sha

More details will be in the docs.